cmds/core/sluinit/uinit.go

```diff
@@ -73,13 +75,38 @@ func main() {
 	}
 	slaunch.Debug("Collectors completed")
 
-	slaunch.Debug("********Step 4: Write eventlog to /boot partition*********")
-	if e := p.EventLog.Persist(); e != nil {
-		log.Printf("EventLog.Persist() failed err=%v", e)
-		os.Exit(1)
+	slaunch.Debug("********Step 4: Measuring target kernel, initrd ********")
+	if err := p.Launcher.MeasureKernel(tpmDev); err != nil {
+		log.Printf("Launcher.MeasureKernel failed err=%v", err)
+		return
 	}
 
-	slaunch.Debug("********Step 5: Launcher called ********")
-	err = p.Launcher.Boot(tpmDev)
-	log.Printf("Boot failed. err=%s", err)
+	slaunch.Debug("********Step 5: Parse eventlogs *********")
+	if err := p.EventLog.Parse(); err != nil {
+		log.Printf("EventLog.Parse() failed err=%v", err)
+		return
+	}
+
+	slaunch.Debug("*****Step 6: Dump logs to disk *******")
+	if err := slaunch.ClearPersistQueue(); err != nil {
+		log.Printf("ClearPersistQueue failed err=%v", err)
+		return
+	}
+
+	slaunch.Debug("********Step *: Unmount all ********")
+	slaunch.UnmountAll()
+
+	slaunch.Debug("********Step 7: Launcher called to Boot ********")
+	if err := p.Launcher.Boot(tpmDev); err != nil {
+		log.Printf("Boot failed. err=%s", err)
+		return
+	}
+}
+
+// unmountAndExit is called on error and unmounts all devices.
+// sluinit ends here.
+func unmountAndExit() {
+	slaunch.UnmountAll()
+	time.Sleep(5 * time.Second) // let queued up debug statements get printed
```

Comment:
We are losing debug statements if we do NOT use sleep....

This has been very painful change for me also. 
Because i hate using sleep. period. 

But our QA team was consistently losing debug statements when sluinit would exit on error . I think Chris even tried to help me debug this, but to this day i cant figure out what race condition is causing sluinit to exit and us losing those debug statements..If we dont use this sleep, customer cant figure out what went wrong in sluinit unless we print the debug statements for them...

I will continue debugging this unless we can fix this.

Reply:
@rminnich This has been very painful change for me also. Because i hate using sleep. period. The issue is we are losing valuable debug statements if we do NOT use sleep at all.

Our QA team was consistently losing debug statements and could not root cause the error when sluinit would abruptly exit on error without printing anything. I think Chris even tried to help me debug this, but to this day i cant figure out what race condition is causing sluinit to exit without printing those debug statements..If we dont use this sleep, customer cant figure out what went wrong in sluinit unless we let sluinit print the debug statements for them...

I do agree that 5 is a random number here which is only going to make things complicated as we move forward.

I would like to create an internal task here so that we keep debugging this as and when we see it on field..